### PR TITLE
retry any failed request regardless of status

### PR DESCRIPTION
**Pull request by bmc-msft**
Simplify the retry logic to always retry any failed request up to 5 times.  The initial delay is now 5 seconds (plus jitter).

Additionally, upon failure we now indicate how many times we failed as well as the timing for each of the request.

### CODE REVIEW

File: src/agent/reqwest-retry/src/lib.rs [hunk 1]
```diff
@@ -195,17 +105,12 @@ mod test {
...
-                2i32,
-                to_backoff_response,
-            )
+            .send_retry(Duration::from_millis(1), 3)
             .await;
 
         if let Err(err) = &resp {
             let as_text = format!("{}", err);
-            assert!(as_text.contains(MAX_RETRY_ERROR_MESSAGE));
+            assert!(as_text.contains("request failed after"));
```
Comment (chkeita) [inline src/agent/reqwest-retry/src/lib.rs:113]: I know this was the existing logic but maybe we need to introduce a RetryError for this case.
not a blocking comment.

**Review by chkeita — APPROVED** [review]

File: src/agent/reqwest-retry/src/lib.rs [hunk 2]
```diff
@@ -1,173 +1,83 @@
...
-use anyhow::Result;
+use anyhow::{format_err, Result};
 use async_trait::async_trait;
 use backoff::{self, future::FutureOperation, ExponentialBackoff};
 use onefuzz_telemetry::warn;
-use reqwest::{Response, StatusCode};
+use reqwest::Response;
 use std::{
-    sync::atomic::{AtomicI32, Ordering},
+    sync::atomic::{AtomicUsize, Ordering},
```
Comment (ranweiler) [inline src/agent/reqwest-retry/src/lib.rs:10]: 👍

File: src/agent/reqwest-retry/src/lib.rs [hunk 3]
```diff
@@ -1,173 +1,83 @@
...
     let result = op
         .retry_notify(
             ExponentialBackoff {
                 current_interval: retry_period,
                 initial_interval: retry_period,
-                max_elapsed_time: Some(max_elapsed_time),
                 ..ExponentialBackoff::default()
             },
-            |err, _| warn!("transient http error: {}", err),
+            |err, dur| warn!("request attempt failed at {:?}: {}", dur, err),
```
Comment (ranweiler) [inline src/agent/reqwest-retry/src/lib.rs:53]:
```suggestion
            |err, dur| warn!("request attempt failed after {:?}: {}", dur, err),
```

**Review by ranweiler — APPROVED** [review]

File: src/agent/reqwest-retry/src/lib.rs (hunk 1 above)
Comment (chkeita) [reply src/agent/reqwest-retry/src/lib.rs:113]: The siggestion here is to have send_retry return a Result<Response, SendRetryError> and move the type of failure inside the SendRetryError. That way we don't have to do string contains

Comment (bmc-msft) [reply src/agent/reqwest-retry/src/lib.rs:113]: That's reasonable, but let's do that as a follow-on PR.

**Pull request merged**